clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreAwarePartitionWiseKafkaConsumerService.java

```diff
@@ -0,0 +1,200 @@
+      final ConsumerPoolType poolType,
+      final PubSubConsumerAdapterFactory consumerFactory,
+      final Properties consumerProperties,
+      final long readCycleDelayMs,
+      final int numOfConsumersPerKafkaCluster,
+      final IngestionThrottler ingestionThrottler,
+      final KafkaClusterBasedRecordThrottler kafkaClusterBasedRecordThrottler,
+      final MetricsRepository metricsRepository,
+      final String kafkaClusterAlias,
+      final long sharedConsumerNonExistingTopicCleanupDelayMS,
+      final TopicExistenceChecker topicExistenceChecker,
+      final boolean liveConfigBasedKafkaThrottlingEnabled,
+      final PubSubMessageDeserializer pubSubDeserializer,
+      final Time time,
+      final AggKafkaConsumerServiceStats stats,
+      final boolean isKafkaConsumerOffsetCollectionEnabled,
+      final ReadOnlyStoreRepository metadataRepository,
+      final boolean isUnregisterMetricForDeletedStoreEnabled) {
+    super(
+        poolType,
+        consumerFactory,
+        consumerProperties,
+        readCycleDelayMs,
+        numOfConsumersPerKafkaCluster,
+        ingestionThrottler,
+        kafkaClusterBasedRecordThrottler,
+        metricsRepository,
+        kafkaClusterAlias,
+        sharedConsumerNonExistingTopicCleanupDelayMS,
+        topicExistenceChecker,
+        liveConfigBasedKafkaThrottlingEnabled,
+        pubSubDeserializer,
+        time,
+        stats,
+        isKafkaConsumerOffsetCollectionEnabled,
+        metadataRepository,
+        isUnregisterMetricForDeletedStoreEnabled);
+    this.LOGGER =
+        LogManager.getLogger(StoreAwarePartitionWiseKafkaConsumerService.class + " [" + kafkaUrlForLogger + "]");
+    this.storeRepository = metadataRepository;
+  }
+
+  @Override
+  protected synchronized SharedKafkaConsumer pickConsumerForPartition(
+      PubSubTopic versionTopic,
+      PubSubTopicPartition topicPartition) {
+    String resourceIdentifier = getResourceIdentifier(versionTopic);
+    Map<Integer, Long> consumerToLoadMap = new VeniceConcurrentHashMap<>();
+    for (int i = 0; i < getConsumerToConsumptionTask().size(); i++) {
+      SharedKafkaConsumer consumer = getConsumerToConsumptionTask().getByIndex(i).getKey();
+      if (topicPartition.getPubSubTopic().isRealTime()
+          && alreadySubscribedRealtimeTopicPartition(consumer, topicPartition)) {
+        getLOGGER().info(
+            "Consumer ID: {} has already subscribed the same real time topic-partition: {}, will assign MAX load to avoid being picked.",
+            i,
+            topicPartition);
+        consumerToLoadMap.put(i, Long.MAX_VALUE);
+        continue;
+      }
+      long baseAssignmentCount = consumer.getAssignmentSize();
+      long storeSubscriptionCount =
+          getResourceIdentifierToConsumerMap().getOrDefault(resourceIdentifier, Collections.emptyMap())
+              .getOrDefault(consumer, 0);
+      long overallLoad = storeSubscriptionCount * DEFAULT_STORE_SUBSCRIPTION_LOAD + baseAssignmentCount;
```

Comment:
Seems the resourceIdentifier only cares about the versionTopic passing here: . For leader and follower partitions, there might be such a case for when leader and follow sharing same pool: 

assume there are 2 consumers, subscription comes in this order (leader rt topic partition 1, follower vt partition 2, leader rt topic partition 3), so ideally leader rt topic partition will get consumer 1 and follower vt partition 2 get consumer 2, but leader rt topic partition 3 will still goes to consumer 1.  Consumer 1 serves 2 leader rt topic partition, but consumer 2 serves 0 leader rt topic partition. So there still some long-tail issue if we want to evenly distribute the leader partition.

Reply:
Also synced offline, we can start with this, as we mostly care about heavy write workload pool's which is AA/WC pool, and for that LEADER has already been picked out.